pkg/upload/dsym.go

```diff
@@ -0,0 +1,160 @@
+package upload
+
+import (
+	"path/filepath"
+
+	"github.com/bugsnag/bugsnag-cli/pkg/ios"
+	"github.com/bugsnag/bugsnag-cli/pkg/log"
+	"github.com/bugsnag/bugsnag-cli/pkg/server"
+	"github.com/bugsnag/bugsnag-cli/pkg/utils"
+)
+
+type Dsym struct {
+	VersionName string      `help:"The version of the application."`
+	Scheme      string      `help:"The name of the scheme to use when building the application."`
+	Dev         bool        `help:"Indicates whether the application is a debug or release build"`
+	DsymPath    string      `help:"Path to the dSYM" type:"path"`
+	Plist       string      `help:"Path to the Info.plist file" type:"path"`
+	ProjectRoot string      `help:"path to remove from the beginning of the filenames in the mapping file" type:"path"`
+	Path        utils.Paths `arg:"" name:"path" help:"Path to directory or file to upload" type:"path" default:"."`
+}
+
+func ProcessDsym(
+	apiKey string,
+	versionName string,
+	scheme string,
+	dev bool,
+	dsymPath string,
+	plistPath string,
+	projectRoot string,
+	paths []string,
+	endpoint string,
+	timeout int,
+	retries int,
+	overwrite bool,
+	dryRun bool,
+) error {
+
+	var buildSettings *ios.XcodeBuildSettings
+	var dsyms *[]*ios.DwarfInfo
+	var plistData *ios.PlistData
+	var uploadOptions map[string]string
+
+	for _, path := range paths {
+		uploadInfo, err := ios.ProcessPathValue(path, projectRoot)
+		if err != nil {
+			return err
+		}
+
+		// If dsymPath is not set explicitly, use uploadInfo to set it (if available)
+		// uploadInfo.DsymPath is set when <path> is recognised as a dsym path
+		if dsymPath == "" && uploadInfo.DsymPath != "" {
+			dsymPath = uploadInfo.DsymPath
+		}
+
+		// If projectRoot is not set explicitly, use uploadInfo to set it
+		if projectRoot == "" {
+			projectRoot = uploadInfo.ProjectRoot
+		}
+
+		// If scheme is set explicitly, check if it exists
+		if scheme != "" {
+			_, err = ios.IsSchemeInPath(path, scheme, projectRoot)
+			if err != nil {
+				return err
+			}
+		} else {
+
+			// Only when the dsym path is not set, try and work out the scheme
+			if dsymPath == "" {
+				// If the scheme is not set explicitly, try to find it
+				scheme, err = ios.GetDefaultScheme(path, projectRoot)
+				if err != nil {
+					return err
+				}
+			}
+		}
+		log.Info("Using scheme: " + utils.DisplayBlankIfEmpty(scheme))
```

Comment:
Move this into the proceeding if-statement as the scheme isn't being "used" if there is a dsym path already.